src/d3a/models/strategy/__init__.py

```diff
@@ -247,65 +260,70 @@ def post_bid(self, market, price, energy):
+        self._bids[market_id].append(bid)
 
-    def add_bid_to_bought(self, bid, market, remove_bid=True):
-        if market not in self._traded_bids:
-            self._traded_bids[market] = []
-        self._traded_bids[market].append(bid)
+    def add_bid_to_bought(self, bid, market_id, remove_bid=True):
+        if market_id not in self._traded_bids:
+            self._traded_bids[market_id] = []
+        self._traded_bids[market_id].append(bid)
         if remove_bid:
-            self.remove_bid_from_pending(bid.id, market)
+            self.remove_bid_from_pending(bid.id, market_id)
 
     def get_traded_bids_from_market(self, market):
-        if market not in self._traded_bids:
+        if market.id not in self._traded_bids:
             return []
         else:
-            return self._traded_bids[market]
+            return self._traded_bids[market.id]
 
-    def are_bids_posted(self, market):
-        if market not in self._bids:
+    def are_bids_posted(self, market_id):
+        if market_id not in self._bids:
             return False
-        return len(self._bids[market]) > 0
+        return len(self._bids[market_id]) > 0
 
     def get_posted_bids(self, market):
-        if market not in self._bids:
+        if market.id not in self._bids:
             return {}
-        return self._bids[market]
+        return self._bids[market.id]
 
     def event_bid_deleted(self, *, market_id, bid):
         assert ConstSettings.IAASettings.MARKET_TYPE is not 1, \
             "Invalid state, cannot receive a bid if single sided market is globally configured."
-        market = self.area.get_future_market_from_id(market_id)
-        assert market is not None
 
         if bid.buyer != self.owner.name:
             return
-        self.remove_bid_from_pending(bid.id, market)
+        self.remove_bid_from_pending(bid.id, market_id)
 
     def event_bid_changed(self, *, market_id, existing_bid, new_bid):
         assert ConstSettings.IAASettings.MARKET_TYPE is not 1, \
             "Invalid state, cannot receive a bid if single sided market is globally configured."
-        market = self.area.get_future_market_from_id(market_id)
         if new_bid.buyer != self.owner.name:
             return
-        self.add_bid_to_posted(market=market, bid=new_bid)
+        self.add_bid_to_posted(market_id, bid=new_bid)
 
     def event_bid_traded(self, *, market_id, bid_trade):
         assert ConstSettings.IAASettings.MARKET_TYPE is not 1, \
             "Invalid state, cannot receive a bid if single sided market is globally configured."
-        market = self.area.get_future_market_from_id(market_id)
-        assert market is not None
 
         if bid_trade.buyer == self.owner.name:
-            self.add_bid_to_bought(bid_trade.offer, self.area.get_future_market_from_id(market_id))
+            self.add_bid_to_bought(bid_trade.offer, market_id)
+
+    def event_market_cycle(self):
+        from d3a.models.const import ConstSettings
```

Comment:
Is this due to circular import or can this be moved to the top?

Reply:
Let me try, I think it is possible to move it to the top, thanks!